helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -202,126 +214,130 @@ public void disconnect() {
-          } else {
-            System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, bucketSize);
-            copyPtr += bucketSize;
-          }
-        }
+    // Compute N - number of buckets
+    int numBuckets = (dataSize + _bucketSize - 1) / _bucketSize;
+    byte[] compressedRecord = new byte[dataSize];
+    String dataPath = path + "/" + versionToRead;
 
-        // Decompress the byte array
-        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(compressedRecord);
-        byte[] serializedRecord;
-        try {
-          serializedRecord = GZipCompressionUtil.uncompress(byteArrayInputStream);
-        } catch (IOException e) {
-          throw new HelixException(String.format("Failed to decompress path: %s!", path), e);
-        }
+    List<String> paths = new ArrayList<>();
+    for (int i = 0; i < numBuckets; i++) {
+      paths.add(dataPath + "/" + i);
+    }
 
-        // Deserialize the record to retrieve the original
-        ZNRecord originalRecord = (ZNRecord) _zkSerializer.deserialize(serializedRecord);
-        return new HelixProperty(originalRecord);
-      } finally {
-        // Critical section for read ends here
-        unlock(path);
+    // Async get
+    List<byte[]> buckets = _zkBaseDataAccessor.get(paths, null, AccessOption.PERSISTENT, true);
+
+    // Combine buckets into one byte array
+    int copyPtr = 0;
+    for (int i = 0; i < numBuckets; i++) {
+      if (i == numBuckets - 1) {
+        // Special treatment for the last bucket
+        System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, dataSize % bucketSize);
+      } else {
+        System.arraycopy(buckets.get(i), 0, compressedRecord, copyPtr, bucketSize);
+        copyPtr += bucketSize;
       }
     }
-    throw new HelixException(String.format("Could not acquire lock for read. Path: %s", path));
+
+    // Decompress the byte array
+    ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(compressedRecord);
+    byte[] serializedRecord;
+    try {
+      serializedRecord = GZipCompressionUtil.uncompress(byteArrayInputStream);
+    } catch (IOException e) {
+      throw new HelixException(String.format("Failed to decompress path: %s!", path), e);
+    }
+
+    // Deserialize the record to retrieve the original
+    ZNRecord originalRecord = (ZNRecord) _zkSerializer.deserialize(serializedRecord);
+    return new HelixProperty(originalRecord);
+  }
+
+  @Override
+  public void close() {
+    disconnect();
   }
 
   /**
-   * Returns the last string element in a split String array by /.
+   * Deletes all versions that are older than the currentVersion given.
    * @param path
-   * @return
+   * @param currentVersion
    */
-  private String extractIdFromPath(String path) {
-    String[] splitPath = path.split("/");
-    return splitPath[splitPath.length - 1];
+  private void deleteStaleVersions(String path, String currentVersion) {
```

Comment:
I think we are not going to apply the read retry if the in-progress read fails because of deleting, right?
In this case, we should still have some protection mechanism to prevent read failure. One way is to delay the delete for 30s, for example. Meaning we always delete the older version if it has been rolled out for more than 30s. In this case, 99.99% readers that reads the old version would be finished for sure.

This could be tricky. We can have a meeting with the team to brainstorm.

Reply:
Retries should be taken care of by the caller. And adding the delay may not be the cleanest solution. At any rate, I feel that that is out of scope of this PR.